examples/server/server.cpp

```diff
@@ -0,0 +1,794 @@
+            is_antiprompt = true;
+            return result;
+          }
+        }
+      }
+      if (n_past > 0)
+      {
+        is_interacting = false;
+      }
+    }
+    if (params.interactive && n_remain <= 0 && params.n_predict != -1)
+    {
+      n_remain = params.n_predict;
+      is_interacting = true;
+    }
+    return result;
+}
+
+std::string Llama::inference() {
+  llama_token tkn = nextToken();
+  if(tkn == -1) {
+    return "";
+  }
+  std::vector<llama_token> tokens_completion;
+  tokens_completion.push_back(tkn);
+
+  // Avoid add the user or assistant tag to the response
+
+  int match_token = 1;
+  if(tokens_completion[0] == user_tag_tokens[0]) {
+    while(true) {
+      if(match_token == user_tag_tokens.size()) { // all user tag tokens matched, return empty inference
+        return "";
+      }
+      tkn = nextToken();
+      tokens_completion.push_back(tkn);
+      if(tkn == user_tag_tokens[match_token]) { // the token follow the sequence
+        match_token++;
+      } else if(match_token < user_tag_tokens.size()) { // no complete all user tag
+        break;
+      }
+    }
+  }
+  if(tokens_completion[0] == assistant_tag_tokens[0]) {
+    bool execute_matching = true;
+    if(tokens_completion.size() > 1) { // if user tag had been tested
+      for(int i = 1;i < assistant_tag_tokens.size(); i++) {
+        if(i >= tokens_completion.size()) {
+          match_token = i;
+          break;
+        }
+        if(tokens_completion[i] == assistant_tag_tokens[i]) {
+          continue;
+        } else {
+          execute_matching = false;
+          break;
+        }
+      }
+    }
+    while(execute_matching) {
+      if(match_token == assistant_tag_tokens.size()) { // all assistant tag tokens matched, return empty inference
+        return "";
+      }
+      tkn = nextToken();
+      tokens_completion.push_back(tkn);
+      if(tkn == assistant_tag_tokens[match_token]) { // the token follow the sequence
+        match_token++;
+      } else if(match_token < assistant_tag_tokens.size()) { // no complete all user tag
+        break;
+      }
+    }
+  }
+  std::string result = "";
+  for(llama_token token : tokens_completion) {
+    result += llama_token_to_str(ctx, token);
+  }
+  return result;
+}
+
+void Llama::release() {
+  // TODO: Clean the context
+  // llama_free(ctx);
+}
+
+void server_print_usage(int /*argc*/, char ** argv, const gpt_params & params) {
+    fprintf(stderr, "usage: %s [options]\n", argv[0]);
+    fprintf(stderr, "\n");
+    fprintf(stderr, "options:\n");
+    fprintf(stderr, "  -h, --help            show this help message and exit\n");
+    fprintf(stderr, "  -s SEED, --seed SEED  RNG seed (default: -1, use random seed for < 0)\n");
+    fprintf(stderr, "  --memory_f32          use f32 instead of f16 for memory key+value\n");
+    fprintf(stderr, "  --keep                number of tokens to keep from the initial prompt (default: %d, -1 = all)\n", params.n_keep);
```

Comment:
`n_keep` should be passed to the API, since it determines how any given input is evaluated.

Reply:
You're right